:sparkles: [automated-actions] action throttling support

chassing opened the pull request:
This PR enhances the OPA authorization to respect the app-interface `maxOps` settings for an action.

These are the most significant changes:

* **OPA** checks the `max_ops` setting by counting the previous user actions (`action-list`)
  * logging for HTTP errors
  * logging exceeded limits
* `action-list` got a `max_age_minutes` option to list actions newer than X minutes
* `OPA` got a `skip_users` setting to avoid additional OPA requests from an OPA API call
* `max_ops: null` means no limits, useful for "non actions" like `me` or `action-list`
* OPA rego tests enhanced and refactored + linter checks on all rego files

Ticket: [APPSRE-12021](https://issues.redhat.com/browse/APPSRE-12021)

patjlm commented on packages/opa/authz/rbac.rego:
```diff
@@ -47,3 +69,75 @@ valid_params(expected, provided) if {
+}
+
+handle_response(api_url, username, response) if {
+	response.status_code == 200
```
nothing between 201 and 299?

chassing replied:
It isn't needed, the automated-action API responds with `200` or an error code.

hemslo commented on packages/automated_actions/automated_actions/db/models/_action.py:
```diff
@@ -93,12 +95,43 @@ def set_final_state(
+                # If no status is provided, but max_age is, return actions
+                # for the user not older than max_age
+                return cls.owner_index.query(
+                    username,
+                    range_key_condition=cls.updated_at
+                    >= int(dt.now(tz=UTC).timestamp() - max_age_val),
```
I checked index in dynamodb, range key is `status`

chassing replied:
I will manually migrate (delete and recreate) the existing tables.

hemslo reviewed (approved):
LGTM!

rporres commented:
We need to make sure that a throttling error is returned in this case. It looks that we're going to return a 403 when a user tries a certain automated action too many times whereas we should be returning a 429.

chassing commented:
@rporres, PTAL. I changed all the discussed things:

* 403 vs 429 http return code
* removed `skip_users` support. All OPA -> automated-actions REST calls are also authN & authZ

rporres reviewed (commented):
minor things, fix at your leisure.

chassing commented on packages/automated_actions/tests/auth/test_opa.py:
```diff
@@ -80,54 +44,40 @@ async def test_opa_user_is_authorized(
-    with pytest.raises(HTTPException):
-        await opa.user_is_authorized(user=user, obj="endpoint", params={"foo": "bar"})
+def test_opa_user_is_authorized(opa: OPA) -> None:
+    opa.user_is_authorized({"authorized": True})
```
The method throws an exception in case the user isn't allowed to.

chassing commented on packages/automated_actions/automated_actions/auth.py:
```diff
@@ -227,58 +227,48 @@ def __init__(
+        if opa_decision.status_code != status.HTTP_200_OK:
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail="OPA returned unexpected http status code",
```
Done in [481e2f3](https://github.com/app-sre/automated-actions/pull/100/commits/481e2f3f8d310bfccf3b6bd077509f7dfba54a67)
